### ISSUE TITLE
Prevent AngleSharp from converting HTML entities to actual characters

### ISSUE THREAD

**Issue by passledevelopers** [AngleSharp/AngleSharp#396]
Is there a way to prevent AngleSharp from converting an HTML entity (for example `&#169;`) to the actual character (ie. `©`)?

thanks,
John.

**Comment by FlorianRappl** [comment]
Currently, there is not - however, I think this question already came along #386 and I see two options:
- Direct access to the source (it is only available internally right now)
- Option to prevent AngleSharp to do the conversion (even though this part is specified in the official parser specification and therefore should be done by any HTML5 conform parser)

The two options are not necessarily exclusive, but an implementation priority needs to be set.

**Comment by passledevelopers** [comment]
Hi Florian,

thank for your quick reply (and all your excellent work on AngleSharp).

My situation is that I'm using [PreMailer.net](https://github.com/milkshakesoftware/PreMailer.Net) which uses AngleSharp under the hood. Unfortunately, the fact that HTML entities are converted makes PreMailer.net practically useless, because major mail clients cannot display the decoded entities. 

I was planning on rolling my own CSS inliner, hence wanting to see if AngleSharp had the option to not decode.

Either of the solutions you have proposed would work in my scenario I think,

thanks again,
John.

**Comment by FlorianRappl** [comment]
I will include this parser option in the HF I will deploy today (v0.9.8.1).

**Comment by passledevelopers** [comment]
That would be fantastic - thanks very much.
John.

**Comment by FlorianRappl** [comment]
Hm, I don't think the option with the original source would work for you. I am not sure about your specific problem, but I think you rely on a serialization of the DOM (and you'll do some DOM manipulation beforehand). The original source is - as the name says - only the original source and won't change with any DOM manipulations.

That being said for the HF1 I am not able to provide option 2 - but option 1 is included (however, see the paragraph above why I don't think option 1 will solve your problem).

The main issue with option 2 is that leaving the entity as-is the serialization of the DOM would change the meaning (leaving `&infin;` to become `&amp;infin;`). Furthermore, the entity is not only used in text, but also in attributes etc. - which excludes using the `Entity` class for transporting such nodes.

TL;DR: Option 1 is included (but potentially not really useful for you) and option 2 is considered in the future, but more time needs to be spent on evaluating potential implementations here.

**Comment by passledevelopers** [comment]
Okay, I appreciate your help on this Florian. I will find another solution.

many thanks,
John.

**Comment by horch004** [comment]
Hi,

We also use PreMailer.Net to move styles inline. And we had the same problem with entities getting converted to their 'real' characters. Since we sometimes still use the iso-8859-1 charset, characters not in the charset should be converted back to entities when serializing to html. It seems the HtmlMarkupFormatter does not take the document charset into account. This might be difficult to do.

Our solution was to use the HtmlAgilityPack.HtmlEntity.Entitize static method to convert characters back to their html entities. But it's a bit unfortune to use two html parser components to do the job.

For a simple test I created my own HtmlMarkupFormatter implementing IMarkupFormatter and redirecting all methods to the AngleSharp HtmlMarkupFormatter.Instance. Only the Text(ICharacterData text) method I changed to `String IMarkupFormatter.Text(ICharacterData text) => Entitize(HtmlMarkupFormatter.Instance.Text(text));` where Entitize is a very simple method which entitizes all character codes above the 127 to the hex representation.

But I took the most simple solution and just use the very simple Entitize function to entitize the PreMailer.Net result.

**Comment by enkelmedia** [comment]
Hi Guys! 

Is there still no build in way of doing this? I wan to replace Html Agility Pack with AngleSharp to manipulate HTML, ie from a Rich Text Editor. But if there is no way to get the original symbols back I'm not sure if it's a good idea. 

Is AngleSharp not intended for manipulation with the source data being kept?

**Comment by FlorianRappl** [comment]
> Is AngleSharp not intended for manipulation with the source data being kept?

You cannot really do that. By the spec HTML5 is not round-trip preserving so you may edit the raw source in a way that it cannot easily ping back.

But I am not sure if I understand your use case. If you parse a source with AngleSharp you can always know what DOM node came from what part of the source. I don't know what else you'd want to know / do there.

Regarding the intention I guess you've read the README. Just let me quote as a reminder:

> The project aims to bring a solid implementation of the W3C DOM for HTML, SVG, MathML, and CSS to the CLR - all written in C#. The idea is that you can basically do everything with the DOM in C# that you can do in JavaScript (plus, of course, more).

**Comment by enkelmedia** [comment]
@FlorianRappl My use case would be ie. processing Rich Text (TinyMCE) input from a user.

Ie. I would like to take a string of HTML and replace all relative img-tags with absolut links. But at the same time I don't want all characters inside a P-tag (that I don't touch) to be replaced/parsed.

I guess that this isn't really the intention of Angle Sharp then

**Comment by FlorianRappl** [comment]
There are options to ignore certain modifications such as the character replacement.

Other than that as mentioned you always have a ref to the source code. So you can even work and manipulate the original source code via such references.

**Comment by enkelmedia** [comment]
Huge thank you @FlorianRappl for your patience with my questions and thanks for getting back.

`"There are options to ignore certain modifications such as the character replacement."` that sounds like what I need any pointers to more info about that?

How would I use the reference to the original source? I haven't seen examples in the docs around this.

**Comment by FlorianRappl** [comment]
Does the `IsNotConsumingCharacterReferences` option help? See https://github.com/AngleSharp/AngleSharp/blob/devel/src/AngleSharp/Html/Parser/HtmlParserOptions.cs#L77

**Comment by jasekiw** [comment]
Hi @FlorianRappl 

I am looking into this issue for Premailer.Net. I am helping out with maintenance and hope to resolve this issue for Premailer.Net users.

I turned on IsNotConsumingCharacterReferences but it did something strange. Is this a bug or is there something I am not understanding here?

```
public class AngleSharpTests
{
	[Fact]
	public void HtmlDocument_ToHtml_ShouldNotEffectHtmlEntities()
	{
		string htmlEncoded = "&lt;&amp;&gt;&nbsp;&copy;";
		string input = $"<html><head></head><body><div>{htmlEncoded}</div></body></html>";
		var document = new HtmlParser(new HtmlParserOptions()
		{
			IsNotConsumingCharacterReferences = true
		}).ParseDocument(input);
		var output = document.ToHtml();
		Assert.Equal<object>(input, output); // use <object> to get full string output
	}
}
```

```
Xunit.Sdk.EqualException
Assert.Equal() Failure
Expected: <html><head></head><body><div>&lt;&amp;&gt;&nbsp;&copy;</div></body></html>
Actual:   <html><head></head><body><div>&amp;lt;&amp;amp;&amp;gt;&amp;nbsp;&amp;copy;</div></body></html>
```

**Comment by FlorianRappl** [comment]
This is not a bug. You have two things:

- the parser. It converts html to an object representation.
- the serializer. It converts the object representation to html.

The option you use changes the behavior of the parser. The serializer still acts unchanged. Hence as the ampersands are kept they are properly serialized.

**Comment by jasekiw** [comment]
@FlorianRappl 

I see what you mean. I noticed I can implement my own IMarkupFormatter. I did some testing an stumbled upon this result:

I implemented `MyFormatter` by following the guide here: https://github.com/AngleSharp/AngleSharp/blob/devel/docs/tutorials/02-Options.md

I inspected how the original formatter did it's work and found the EscapeText method and implemented my own version to convert the copyright symbol to the html entity,
```
public class MyFormatter: IMarkupFormatter
{
	public string CloseTag(IElement element, bool selfClosing) => HtmlMarkupFormatter.Instance.CloseTag(element, selfClosing);
	public string Comment(IComment comment) => HtmlMarkupFormatter.Instance.Comment(comment);
	public string Doctype(IDocumentType doctype) => HtmlMarkupFormatter.Instance.Doctype(doctype);
	public string LiteralText(ICharacterData text) => HtmlMarkupFormatter.Instance.LiteralText(text);
	public string OpenTag(IElement element, bool selfClosing) => HtmlMarkupFormatter.Instance.OpenTag(element, selfClosing);
	public string Processing(IProcessingInstruction processing) => HtmlMarkupFormatter.Instance.Processing(processing);

	public string Text(ICharacterData text)
	{
		var content = text.Data;
		return EscapeText(content);
	}
	public const Char CopyRightSymbol = (Char)0x00A9;
	
	/// <summary>
	/// Escapes the given text by replacing special characters with their
	/// HTML entity (amp, nobsp, lt, and gt).
	/// </summary>
	/// <param name="content">The string to alter.</param>
	/// <returns>The altered string.</returns>
	public static String EscapeText(String content)
	{
		var temp = StringBuilderPool.Obtain();

		for (var i = 0; i < content.Length; i++)
		{
			switch (content[i])
			{
				case Symbols.Ampersand: temp.Append("&amp;"); break;
				case Symbols.NoBreakSpace: temp.Append("&nbsp;"); break;
				case Symbols.GreaterThan: temp.Append("&gt;"); break;
				case Symbols.LessThan: temp.Append("&lt;"); break;
				case CopyRightSymbol: temp.Append("&copy;"); break;
				default: temp.Append(content[i]); break;
			}
		}

		return temp.ToPool();
	}
}
```

I then used the formatter in my test and it just worked the way I wanted it to.


```
	[Fact]
	public void HtmlDocument_ToHtml_ShouldNotEffectHtmlEntities()
	{
		string htmlEncoded = "&lt;&amp;&gt;&nbsp;&copy;";
		string input = $"<html><head></head><body><div>{htmlEncoded}</div></body></html>";
		var document = new HtmlParser().ParseDocument(input);
		var output = document.ToHtml(new MyFormatter());
		Assert.Equal<object>(input, output); // use object to get full string output
	}
```


Output is the same as the input which is what I expected. I wanted to ask, do you think this is a fine solution to the problem? One pitfall I see is I won't receive updates to the EscapeText method when AngleSharp decides to update it. I have a request to keep all entities from being converted, so in order to support that I would need to convert all symbols back to html entities, not just the copyright symbol. I see AngleSharp has a IEntityProvider, I would basically need the inverse of that provider.

**Comment by jasekiw** [comment]
@enkelmedia I think for your use case you would most likely want to do the following:

Setup your own formatter to output literal text instead of escaping:

```
class MyFormatter : IMarkupFormatter
{
    public string CloseTag(IElement element, bool selfClosing) => HtmlMarkupFormatter.Instance.CloseTag(element, selfClosing);
    public string Comment(IComment comment) => HtmlMarkupFormatter.Instance.Comment(comment);
    public string Doctype(IDocumentType doctype) => HtmlMarkupFormatter.Instance.Doctype(doctype);
    public string LiteralText(ICharacterData text) => HtmlMarkupFormatter.Instance.LiteralText(text);
    public string OpenTag(IElement element, bool selfClosing) => HtmlMarkupFormatter.Instance.OpenTag(element, selfClosing);
    public string Processing(IProcessingInstruction processing) => HtmlMarkupFormatter.Instance.Processing(processing);
    public string Text(ICharacterData text) => HtmlMarkupFormatter.Instance.LiteralText(text);
}
```
Turn on `IsNotConsumingCharacterReferences`, and use your formatter.
```
var document = new HtmlParser(new HtmlParserOptions()
{
	IsNotConsumingCharacterReferences = true
}).ParseDocument(input);
var output = document.ToHtml(new MyFormatter());
```

This should prevent any html entities from getting converted.

**Comment by FlorianRappl** [comment]
> I wanted to ask, do you think this is a fine solution to the problem?

I'd say yes - this is the way.

> One pitfall I see is I won't receive updates to the EscapeText method when AngleSharp decides to update it.

Hm the only reason for updating it is when the spec changes - and this part of the spec seems so stable that I doubt we'll ever update it (but, of course, there is always a chance of this happening...).

> I see AngleSharp has a IEntityProvider, I would basically need the inverse of that provider.

Maybe we can introduce something like this. Maybe the provider can do both ways?

**Comment by jasekiw** [comment]
@FlorianRappl 

> Maybe we can introduce something like this. Maybe the provider can do both ways?

That would be great! Let me know if you need any help.

**Comment by jasekiw** [comment]
@FlorianRappl Any update on whether you can introduce a bidirectional IEntityProvider?

**Comment by FlorianRappl** [comment]
Well (a) it's OSS so anyone can provide a PR and (b) have you checked the latest release? It's already out.

**Comment by jasekiw** [comment]
@FlorianRappl 

First, I really appreciate you implementing this!

> (a) it's OSS so anyone can provide a PR

I am aware, this is why I asked if you needed any help. (insinuating I could provide a PR) - maybe a I could have been more direct about this.

> (b) have you checked the latest release? It's already out.

I did not know that you decided to move forward with this. I assumed there would be some sort of update on this thread if so. Maybe I should have created a separate issue to track it? I will be sure to do that in the future.

This will be really helpful for Premailer.Net. Thank you again for taking your time to implement this.

**Comment by enkelmedia** [comment]
Great tone of voice @jasekiw! @FlorianRappl I understand that OSS might be stressful but if you need help, OSS is all about asking for it. Than you both for everything that you done for the .NET Community!
